src/blueteam/Gardener.java

```diff
@@ -137,30 +165,137 @@ boolean build(RobotType type) {
+		return false;
+	}
 
+	/**
+	 * Checks whether there is a lot (const) trees around
+	 *
+	 * @return true if is more than TeamConstants.GARDENER_NUM_OF_TREES_TO_BUILD_LUMBER
+	 */
+	boolean isInWoods() {
+		TreeInfo[] nearbyTrees = rc.senseNearbyTrees(TeamConstants.GARDENERS_DEFAULT_FREE_SPOT_RADIUS, Team.NEUTRAL);
+		System.out.println(nearbyTrees.length);
+		if (nearbyTrees.length > TeamConstants.GARDENER_NUM_OF_TREES_TO_BUILD_LUMBER)
+			return true;
+		else
+			return false;
+	}
+	@Override void dodge()
+	{
+
+	}
+	/**
+	 * States:
+	 * <p>
+	 * STARTING <-> ONLYSOLDIER
+	 * \
+	 * \-----> FINDING <-> LETSCHOP
+	 * \
+	 * \-----> BUILDING
+	 */
+	@Override void step() {
+		roundCounter++;
 		switch (state) {
+		case STARTING:
+			currentDir = randomDirection();
+			if (isEnemyArchonNear()) {
+				state = GardenerState.ONLYSOLDIERS;
+			} else
+				state = GardenerState.FINDING;
+			break;
 		case FINDING:
-			if (findSpot()) {
-				state = GardenerState.BUILDING;
-				System.out.println("Gardener ID: " + rc.getID() + " found place to build garden");
-			} else {
-				tryMove(randomDirection());
+			while (roundCounter < 20) {
+				if (findSpot())
+					state = GardenerState.BUILDING;
+				roundCounter++;
+				Clock.yield();
 			}
+			if (isInWoods() && lumberjackBuilded < 1) {
+				state = GardenerState.LETSCHOP;
+				lumberjackBuilded++;
+			} else if (findSpot())
+				state = GardenerState.BUILDING;
 			break;
 		case BUILDING:
-			buildGarden();
+			// try to build one tree in garden
+			if (buildGarden())
+				initTreeNum++;
+			// water garden
 			waterGarden();
-			//TODO decide according to game state whether to build lumberjacks,
-			//soldiers or tanks. Logic how many gardeners should be build should be in archon
-			double rnd = Math.random();
-			if (rnd < 0.1)
-				build(RobotType.LUMBERJACK);
-			else if (rnd < 0.9)
+			//build atleast 3 trees
+			if (initTreeNum < 3)
+				break;
+			// force to build robots
+			if (!scoutBuilded && getRobotCount(RobotType.SCOUT) == 0) {
+				while (!build(RobotType.SCOUT)) {
+					waterGarden();
+					Clock.yield();
+				}
+				scoutBuilded = true;
+			} else if (!soldierBuilded) {
+				while (!build(RobotType.SOLDIER)) {
+					waterGarden();
+					Clock.yield();
+				}
+				soldierBuilded = true;
+			}
+			if (filterByType(rc.senseNearbyRobots(TeamConstants.GARDENERS_DEFAULT_FREE_SPOT_RADIUS, enemy),
+					RobotType.SOLDIER).size() > 0)
+				build(RobotType.SOLDIER);
+			// try to build again
+			double rnd = rand.nextDouble();
+			if (rnd < 0.25) {
+				if (getRobotCount(RobotType.SCOUT) < 2)
+					build(RobotType.SCOUT);
+				else if (getRobotCount(RobotType.SOLDIER) < 40)
+					build(RobotType.SOLDIER);
+			} else if (rnd < 0.5) {
+				if (getRobotCount(RobotType.LUMBERJACK) < 3)
+					build(RobotType.LUMBERJACK);
+				else if (getRobotCount(RobotType.SOLDIER) < 40)
+					build(RobotType.SOLDIER);
+			} else {
+				if (getRobotCount(RobotType.SOLDIER) < 40)
+					build(RobotType.SOLDIER);
```

Comment:
Consider adding these constatnts to the TeamConstants file. I know that there are some compilation issues with changing these values, but we need to resolve that in that class.